README.md
# Oh My Fast Postgres!

`ohmyfpg` is a Postgres client library for Python that aims to return data as columns. This is often needed when working with numerical data. Usually this is achieved by taking the output of the client library and then convert it either into `numpy` arrays or `pandas` dataframes. When dealing with a large amount of data, this conversion is not much performant.

The goal of this library is to return data already as `numpy` arrays without sacrificing performance. The section of "Performance comparison" goes more in-depth on this topic.

In order to squeeze performance the underlying implementation is written in Rust. The Python layer on top is very thin.

## Why `ohmyfpg`?

When working with Postgres at work we faced multiple times performance issues. Most of the times our reactions were along the lines of: "OMG", "F*****g PG", etc. So `ohmyfpg` is kinda a mix of the two, but where the `f` now stands for `fast`.

(To be fair, when we faced performance issues with Postgres was most of the times because of our inexperience with tuning the server configurations.)

## Installation

```
pip install ohmyfpg
```

## Quickstart

```
import asyncio
import ohmyfpg

DSN = 'postgres://postgres:postgres@postgres:5432/postgres'
QUERY = 'SELECT * FROM performance_test'

async def main():
    conn = await ohmyfpg.connect(DSN)
    print(await conn.fetch(QUERY))

if __name__ == '__main__':
    asyncio.run(main())
```

## Performance comparison

The image below compares the performance of `ohmyfpg`, `asyncpg`, and `psycopg`. The 6 bars have the following meaning:
- `ohmyfpg`: plain fetch,
- `asyncpg`: plain fetch,
- `psycopg`: plain fetch,
- `ohmyfpg-pandas`: plain fetch + conversion to `pandas` Dataframe,
- `asyncpg-pandas`: plain fetch + conversion to `pandas` Dataframe,
- `psycopg-pandas`: plain fetch + conversion to `pandas` Dataframe,

See details [here](performance/compare.py), especially how the conversion to `pandas` Dataframe has been implemented.

The query is a `SELECT *` that has been run on a table with 1mln rows and the following schema:
```
(id INT, foo_bar_int2 INT2, foo_bar_int4 INT4, foo_bar_int8 INT8, foo_bar_float4 FLOAT4, foo_bar_float8 FLOAT8)
```

It has been run inside docker with 8 CPU and 8GB of RAM allocated to the daemon on a MBP with 2.2 GHz 6-Core Intel Core i7 and 16GB 2400 MHz DDR4.

![Performance comparison](performance/performance-comparison.png)

### Detailed summary

#### Plain fetch

- `ohmyfpg` vs. `asyncpg` => **38.4%** (or **1.6x**) faster
- `ohmyfpg` vs. `psycopg` => **58.0%** (or **2.4x**) faster

```
--------------------------------------------------
ohmyfpg
avg: 856.2ms
min: 747ms
p25: 782.25ms
median: 819.0ms
p75: 871.0ms
max: 1335ms
--------------------------------------------------
asyncpg
avg: 1375.3ms
min: 1136ms
p25: 1261.75ms
median: 1330.5ms
p75: 1406.25ms
max: 1925ms
--------------------------------------------------
psycopg
avg: 2023.7ms
min: 1777ms
p25: 1886.75ms
median: 1951.0ms
p75: 2073.0ms
max: 3039ms
--------------------------------------------------
```

#### Plain fetch + conversion to `pandas` Dataframe

- `ohmyfpg-pandas` vs. `asyncpg-pandas` => **64.7%** (or **2.8x**) faster
- `ohmyfpg-pandas` vs. `psycopg-pandas` => **69.8%** (or **3.3x**) faster

```
--------------------------------------------------
ohmyfpg-pandas
avg: 970.6666666666666ms
min: 852ms
p25: 924.25ms
median: 948.0ms
p75: 988.5ms
max: 1292ms
--------------------------------------------------
asyncpg-pandas
avg: 2754.1666666666665ms
min: 2569ms
p25: 2642.75ms
median: 2683.0ms
p75: 2737.75ms
max: 4044ms
--------------------------------------------------
psycopg-pandas
avg: 3193.6ms
min: 2945ms
p25: 3067.0ms
median: 3141.0ms
p75: 3236.75ms
max: 4040ms
--------------------------------------------------
```

## Limitations

This library is highly experimental and has many limitations:
- no support for `NULL`s with unpredictable outcome,
- no support for non-numerical types,
- limited support for authentication,
- no proper logging,
- no support for insert operations,
- no parameters support for prepared statements,
- etc.

## Development

<details>
<summary>Expand</summary>

### How to run the performance comparison

```
docker compose build script
docker compose up -d postgres
docker compose exec -- postgres psql -U postgres

CREATE TABLE performance_test (id INT, foo_bar_int2 INT2, foo_bar_int4 INT4, foo_bar_int8 INT8, foo_bar_float4 FLOAT4, foo_bar_float8 FLOAT8);
INSERT INTO performance_test (
    id,
    foo_bar_int2,
    foo_bar_int4,
    foo_bar_int8,
    foo_bar_float4,
    foo_bar_float8
) VALUES (
    generate_series(1, 1000000),
    trunc(random() * (2*32768) - 32768),
    trunc(random() * (2*2147483648) - 2147483648),
    trunc(random() * (2*9223372036854775808) - 9223372036854775808),
    trunc(random()),
    trunc(random())
);


docker compose up script
docker compose cp script:/usr/src/app/performance-comparison.png ./performance
```

### How to do basic benchmarking

```
docker run -p 5432:5432 --name rust-postgres -e POSTGRES_PASSWORD=postgres -d postgres -c log_min_messages=DEBUG5
```

Data preparation:
```
CREATE TABLE performance_test (id INT, foo_bar_int2 INT2, foo_bar_int4 INT4, foo_bar_int8 INT8, foo_bar_float4 FLOAT4, foo_bar_float8 FLOAT8);
INSERT INTO performance_test (
    id,
    foo_bar_int2,
    foo_bar_int4,
    foo_bar_int8,
    foo_bar_float4,
    foo_bar_float8
) VALUES (
    generate_series(1, 1000000),
    trunc(random() * (2*32768) - 32768),
    trunc(random() * (2*2147483648) - 2147483648),
    trunc(random() * (2*9223372036854775808) - 9223372036854775808),
    trunc(random()),
    trunc(random())
);
```

```
maturin develop --release --manifest-path ohmyfpg/Cargo.toml
python python/examples/simple_query.py
```

```
RUST_BACKTRACE=1 cargo run -r -p ohmyfpg_core --example simple_query
```

### How to do basic profiling

```
sudo CARGO_PROFILE_BENCH_DEBUG=true RUST_BACKTRACE=1 cargo flamegraph -p ohmyfpg_core --example simple_query
```

```
CARGO_PROFILE_BENCH_DEBUG=true RUST_BACKTRACE=1 cargo instruments --release -p ohmyfpg_core --example simple_query -t time
```

```
CARGO_PROFILE_BENCH_DEBUG=true RUST_BACKTRACE=1 cargo instruments --release -p ohmyfpg_core --example simple_query -t Allocations
```
</details>


HISTORY.md
# HISTORY

## Unreleased

### Added

- Added comparison against `psycopg`
- Added initial support to extended query with binary format

### Changed

- Reduced `malloc`s by ~55% for `simple_query` example
- [internal] Replace `Vec<u8>` with `bytes` in the hottest places

## [v0.3.0 - 2022-09-29](https://github.com/se7entyse7en/ohmyfpg/compare/v0.2.1...v0.3.0)

### Changed

- Improved performance and added proper comparison against `asyncpg`

## [v0.2.1 - 2022-09-11](https://github.com/se7entyse7en/ohmyfpg/compare/v0.2.0...v0.2.1)

### Fixed

- Fix version bump of the `Cargo.lock`s files

## [v0.2.0 - 2022-09-11](https://github.com/se7entyse7en/ohmyfpg/compare/v0.1.0...v0.2.0) [YANKED]

### Added

- Add script to compare performance with `asyncpg`
- [internal] Add `RawBackendMessage` in order to be able to split backend message identification from the full payload parsing
- [internal] Add `Framer` that handles eager reading from tcp stream concurrently

### Changed

- Renamed `PyInvalidDSNError` -> `PyInvalidDsnError`
- Improved performance by reading buffer eagerly instead of two syscalls (header + body) per message
- [internal] Split rust part into two workspaces (core + binding) to ease benchmarking and profiling of core

## [v0.1.0 - 2022-08-25](https://github.com/se7entyse7en/ohmyfpg/compare/v0.0.0...v0.1.0)

### Added

- Added a `connect` function returning a `Connection` object
- Implement a `fetch` method returning columnar data as `numpy` arrays with support to numerical Postgres types (`int2`, `int4`, `int8`, `float4`, `float8`)

## [v0.0.0 - 2022-06-26](https://github.com/se7entyse7en/ohmyfpg/compare/95f47c4cee38fad74a969ec34e5169c6e4e23c38...v0.0.0)

- Project inception! :tada:


performance/compare.py
import asyncio
import sys
import psycopg
import time
import numpy as np
import asyncpg
import ohmyfpg
import pandas as pd
import matplotlib.pyplot as plt


TEST_RUNS = 30
HEAT_RUNS = 1
DSN = 'postgres://postgres:postgres@postgres:5432/postgres'
QUERY = 'SELECT * FROM performance_test'


class Runner:

    name = ''

    def __init__(self, heat_runs, runs):
        self._heat_runs = heat_runs
        self._runs = runs

    async def connect(self, dsn):
        raise NotImplementedError()

    async def heat(self, conn, query):
        raise NotImplementedError()

    async def execute(self, conn, query):
        raise NotImplementedError()

    async def run(self, dsn, query):
        print('Connecting...')
        conn = await self.connect(dsn)
        print('Heating...')
        for i in range(self._heat_runs):
            await self.heat(conn, query)

        times = []
        print('Executing...')
        for i in range(self._runs):
            start = time.time()
            await self.execute(conn, query)
            end = time.time()
            times.append(round((end - start) * 1000))

        return (
            times,
            np.mean(times),
            np.min(times),
            np.percentile(times, 25),
            np.median(times),
            np.percentile(times, 75),
            np.max(times),
        )


class OhmyfpgRunner(Runner):

    name = 'ohmyfpg'

    async def connect(self, dsn):
        return await ohmyfpg.connect(dsn)

    async def heat(self, conn, query):
        await conn.fetch(query)

    async def execute(self, conn, query):
        return await conn.fetch(query)


class AsyncpgRunner(Runner):

    name = 'asyncpg'

    async def connect(self, dsn):
        return await asyncpg.connect(dsn, statement_cache_size=0)

    async def heat(self, conn, query):
        await conn.fetch(query)

    async def execute(self, conn, query):
        stmt = await conn.prepare(query)
        return await stmt.fetch()


class PsycopgRunner(Runner):

    name = 'psycopg'

    async def connect(self, dsn):
        return await psycopg.AsyncConnection.connect(dsn)

    async def heat(self, conn, query):
        cursor = conn.cursor()
        await cursor.execute(query)
        await cursor.fetchall()

    async def execute(self, conn, query):
        cursor = conn.cursor()
        await cursor.execute(query)
        return await cursor.fetchall()


class OhmyfpgPandasRunner(OhmyfpgRunner):

    name = 'ohmyfpg-pandas'

    async def execute(self, conn, query):
        await conn.fetch(query)

    async def execute(self, conn, query):
        return pd.DataFrame(await super().execute(conn, query))


class AsyncpgPandasRunner(AsyncpgRunner):

    name = 'asyncpg-pandas'

    async def execute(self, conn, query):
        stmt = await conn.prepare(query)
        columns = [c.name for c in stmt.get_attributes()]
        data = await stmt.fetch()
        return pd.DataFrame(data, columns=columns)


class PsycopgPandasRunner(PsycopgRunner):

    name = 'psycopg-pandas'

    async def execute(self, conn, query):
        cursor = conn.cursor()
        await cursor.execute(query)
        columns = [c.name for c in cursor.description]
        data = await cursor.fetchall()
        return pd.DataFrame(data=data, columns=columns)


async def main():
    runners = [
        OhmyfpgRunner(HEAT_RUNS, TEST_RUNS),
        AsyncpgRunner(HEAT_RUNS, TEST_RUNS),
        PsycopgRunner(HEAT_RUNS, TEST_RUNS),
        OhmyfpgPandasRunner(HEAT_RUNS, TEST_RUNS),
        AsyncpgPandasRunner(HEAT_RUNS, TEST_RUNS),
        PsycopgPandasRunner(HEAT_RUNS, TEST_RUNS),
    ]
    results = {}
    for r in runners:
        print(f'Running test for {r.name}...', file=sys.stderr)
        results[r.name] = await r.run(DSN, QUERY)

    boxplot_data = []
    boxplot_labels = []
    for name, stats in results.items():
        boxplot_data.append(stats[0])
        boxplot_labels.append(name)

        print('-' * 50)
        print(name)
        print(f'avg: {stats[1]}ms')
        print(f'min: {stats[2]}ms')
        print(f'p25: {stats[3]}ms')
        print(f'median: {stats[4]}ms')
        print(f'p75: {stats[5]}ms')
        print(f'max: {stats[6]}ms')
        print('-' * 50)

    fig, ax = plt.subplots(figsize=(12, 10))
    ax.set_title('Performance comparison (ms)', fontsize=18)
    ax.boxplot(boxplot_data, labels=boxplot_labels)
    ax.tick_params(axis='both', which='major', labelsize=12)
    plt.savefig('performance-comparison.png')


if __name__ == '__main__':
    asyncio.run(main())
